File: 31-next-permutation/31-next-permutation.py
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        end = len(nums) - 1
        pos = -10
        for i in range(end, -1, -1):
            if i == end:
                prev = nums[end]
            elif prev <= nums[i]:
                prev = nums[i]
            else:
                j = i + 1
                while(j <= end and nums[j] > nums[i]):
                    j = j + 1
                if j<= end and nums[j] > nums[i]:
                    pos = j
                else:
                    pos = j - 1
                break
        if pos == -10:
            nums.sort()
        else:
            nums[i], nums[pos] = nums[pos], nums[i]
            nums[i+1:] = sorted(nums[i+1:])
        
                
            
                
        

File: 31-next-permutation/31-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int[] rightMax = new int[nums.length];
        LinkedList<Integer> stack = new LinkedList<>();
        for(int i = 0; i < nums.length; i++)
            for(int j = i + 1; j < nums.length; j++){
                if(nums[j] > nums[i]) rightMax[i] = Math.max(rightMax[i], j);        
            }
        int rightMost = -1;
        for(int j = nums.length - 1; j > - 1; j--){
            if(rightMax[j] != 0){
                rightMost = Math.max(rightMost, j);
            }
        }
        if (rightMost != -1){
            int temp = nums[rightMost];
            nums[rightMost] = nums[rightMax[rightMost]];
            nums[rightMax[rightMost]] = temp;
            Arrays.sort(nums, rightMost+1, nums.length);
        }else{
            reverse(nums);
        }
    }
    public void reverse(int[] nums){
        for(int i = 0; i < nums.length / 2; i++){
            int temp = nums[i];
            nums[i] = nums[nums.length - i - 1];
            nums[nums.length - i - 1] = temp;
        }
    }
}